qiskit/pulse/schedule.py

```diff
@@ -1222,24 +1402,154 @@ def assign_parameters(
 
         Returns:
             Schedule with updated parameters.
+
+        Raises:
+            PulseError: When the block is nested into another block.
         """
-        return self._parameter_manager.assign_parameters(
-            pulse_program=self, value_dict=value_dict, inplace=inplace
-        )
+        if not inplace:
+            new_schedule = copy.deepcopy(self)
+            return new_schedule.assign_parameters(value_dict, inplace=True)
 
-    def get_parameters(self, parameter_name: str) -> List[Parameter]:
+        # Update parameters in the current scope
+        self._parameter_manager.assign_parameters(pulse_program=self, value_dict=value_dict)
+
+        for subroutine in self._reference_manager.values():
+            # Also assigning parameters to the references associated with self.
+            # Note that references are always stored in the root program.
+            # So calling assign_parameters from nested block doesn't update references.
+            if subroutine is None:
+                continue
+            subroutine.assign_parameters(value_dict=value_dict, inplace=True)
+
+        return self
+
+    def assign_references(
+        self,
+        subroutine_dict: Dict[str, "ScheduleBlock"],
+        inplace: bool = True,
+    ) -> "ScheduleBlock":
+        """Assign schedule to reference.
+
+        It is only capable of assigning a substantial program to child subroutines
```

Comment:
I don't understand what is meant by `substantial program`?

Reply:
replaced with schedule block. "substantial" is used as an opposite of "opaque" because reference is definition-less instruction.